File: 1D DP/000 Pracitce/Coin Change II/main.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution{
    public:
    int coinchange(const vector<int>& arr, int target){
        int n = arr.size();

        vector<int> dp(target+1);
        dp[0] = 1;

        for (int i=0; i<n; i++){
            for (int j=arr[i]; j < target+1; j++){
                dp[j] = dp[j-arr[i]] + dp[j];
            }
        }

        return dp[target];
    }
};

int main(){

    Solution solution;

    vector<int> coins = {1,2,3};
    int sum = 4;
    cout << solution.coinchange(coins, sum);

    return 0;
}

File: 1D DP/000 Pracitce/Coin Change II/main.py
class Solution:
    def coinchange(self, arr: list[int], target: int):
        
        dp = [0] * (target + 1)
        dp[0] = 1

        for num in arr:
            for i in range(num, target+1):
                dp[i]  = dp[i - num] + dp[i]

        return dp[target]

solution = Solution()
print(solution.coinchange(
    arr = [1,2,3],
    target=4
))